### PR TITLE
Vending TGUI

### PR DESCRIPTION
**Pull request by ghost**
:cl: More Robust Than You
tweak: Vending machines now use TGUI
/:cl:

### CODE REVIEW

**Comment by Cyberboss** [comment]
Because TGUI is the standard

**Comment by MoreRobustThanYou** [comment]
Want pictures? look at a vending machine on hippie since i used their code

**Comment by praisenarsie** [comment]
don't be lazy

**Comment by MoreRobustThanYou** [comment]
capture

File: code/game/machinery/vending.dm [hunk 1]
```diff
@@ -605,6 +436,108 @@
...
+					return
+				if(coin.string_attached)
+					if(prob(50))
+						if(usr.put_in_hands(coin))
+							to_chat(usr, "<span class='notice'>You successfully pull [coin] out before [src] could swallow it.</span>")
+							coin = null
+						else
+							to_chat(usr, "<span class='warning'>You couldn't pull [coin] out because your hands are full!</span>")
+							qdel(coin)
+							coin = null
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: QDEL_NULL

File: code/game/machinery/vending.dm [hunk 2]
```diff
@@ -605,6 +436,108 @@
...
+					else
+						to_chat(usr, "<span class='warning'>You weren't able to pull [coin] out fast enough, the machine ate it, string and all!</span>")
+						qdel(coin)
+						coin = null
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: QDEL_NULL

File: code/game/machinery/vending.dm [hunk 3]
```diff
@@ -605,6 +436,108 @@
...
+				else
+					qdel(coin)
+					coin = null
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: QDEL_NULL

File: code/game/machinery/vending.dm [hunk 4]
```diff
@@ -605,6 +436,108 @@
...
+			else if (!(R in display_records))
+				message_admins("Vending machine exploit attempted by [key_name(usr, usr.client)]!")
+				return
+
+			if (R.amount <= 0)
+				to_chat(usr, "<span class='warning'>Sold out.</span>")
+				return
+			else
+				R.amount--
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: Dangling else

File: code/game/machinery/vending.dm [hunk 5]
```diff
@@ -605,6 +436,108 @@
...
+					else
+						to_chat(usr, "<span class='warning'>You weren't able to pull [coin] out fast enough, the machine ate it, string and all!</span>")
+						qdel(coin)
+						coin = null
+				else
+					qdel(coin)
+					coin = null
+			else if (!(R in display_records))
+				message_admins("Vending machine exploit attempted by [key_name(usr, usr.client)]!")
+				return
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: We don't notify admins of stuff we can prevent like this

File: code/game/machinery/vending.dm [hunk 6]
```diff
@@ -605,6 +436,108 @@
...
+			if (R.amount <= 0)
+				to_chat(usr, "<span class='warning'>Sold out.</span>")
+				return
+			else
+				R.amount--
+
+			use_power(5)
+			if(icon_vend) //Show the vending animation if needed
+				flick(icon_vend,src)
+			new R.product_path(get_turf(src))
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: drop_location()

File: code/game/machinery/vending.dm [hunk 7]
```diff
@@ -605,6 +436,108 @@
...
+			return TRUE
+		if("eject")
+			if(!coin)
+				to_chat(usr, "<span class='notice'>There is no coin in this machine.</span>")
+				return
+			coin.forceMove(loc)
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: drop_location()

File: code/game/machinery/vending.dm [hunk 8]
```diff
@@ -605,6 +436,108 @@
...
+			if(!usr.get_active_held_item())
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: Remove this check

File: code/game/machinery/vending.dm [hunk 9]
```diff
@@ -605,6 +436,108 @@
...
+			var/mob/living/silicon/robot/R = usr
+			if(!(R.module && istype(R.module,/obj/item/robot_module/butler) ))
+				to_chat("<span class='notice'>The vending machine refuses to interface with you, as you are not in its target demographic!</span>")
+				return
+		else
+			to_chat(usr, "<span class='notice'>The vending machine refuses to interface with you, as you are not in its target demographic!</span>")
+			return
+	switch(action)
+		if("vend")
+			if((!allowed(usr)) && !emagged && scan_id)	//For SECURE VENDING MACHINES YEAH
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: Too many parenthesis

File: code/game/machinery/vending.dm [hunk 10]
```diff
@@ -605,6 +436,108 @@
...
+/obj/machinery/vending/ui_act(action, params)
+	if(..())
+		return
+	if(!vend_ready)
+		return
+	if(issilicon(usr))
+		if(iscyborg(usr))
+			var/mob/living/silicon/robot/R = usr
+			if(!(R.module && istype(R.module,/obj/item/robot_module/butler) ))
+				to_chat("<span class='notice'>The vending machine refuses to interface with you, as you are not in its target demographic!</span>")
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: [src] refuses...

File: code/game/machinery/vending.dm [hunk 11]
```diff
@@ -605,6 +436,108 @@
...
+				return
+		else
+			to_chat(usr, "<span class='notice'>The vending machine refuses to interface with you, as you are not in its target demographic!</span>")
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: ditto

File: code/game/machinery/vending.dm [hunk 12]
```diff
@@ -133,6 +135,9 @@
 		stat |= BROKEN
 		icon_state = "[initial(icon_state)]-broken"
 
+/obj/machinery/vending/proc/enable_vend()
+	vend_ready = TRUE
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: Remove this proc

File: code/game/machinery/vending.dm [hunk 13]
```diff
@@ -60,6 +60,8 @@
 	var/obj/item/vending_refill/refill_canister = null		//The type of refill canisters used by this machine.
 	var/refill_count = 3		//The number of canisters the vending machine uses
 
+	var/list/display_records = list()
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: Remove this

File: code/game/machinery/vending.dm [hunk 14]
```diff
@@ -605,6 +436,108 @@
...
+			to_chat(usr, "<span class='notice'>The vending machine refuses to interface with you, as you are not in its target demographic!</span>")
+			return
+	switch(action)
+		if("vend")
+			if((!allowed(usr)) && !emagged && scan_id)	//For SECURE VENDING MACHINES YEAH
+				to_chat(usr, "<span class='warning'>Access denied.</span>")
+				flick(icon_deny,src)
+				return
+			vend_ready = FALSE
+			addtimer(CALLBACK(src, .proc/enable_vend), 10)
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: use VARSET_CALLBACK

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/machinery/vending.dm [hunk 15]
```diff
@@ -475,21 +470,22 @@
...
 			var/mob/living/silicon/robot/R = usr
 			if(!(R.module && istype(R.module,/obj/item/robot_module/butler) ))
-				to_chat("<span class='notice'>The vending machine refuses to interface with you, as you are not in its target demographic!</span>")
+				to_chat("<span class='notice'>[src] refuses to interface with you, as you are not in its target demographic!</span>")
 				return
 		else
-			to_chat(usr, "<span class='notice'>The vending machine refuses to interface with you, as you are not in its target demographic!</span>")
+			to_chat(usr, "<span class='notice'>[src] refuses to interface with you, as you are not in its target demographic!</span>")
 			return
+	var/list/display_records = product_records
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: unused var

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by WJohn** [comment]
Neat!

File: code/game/machinery/vending.dm [hunk 16]
```diff
@@ -605,6 +431,102 @@
...
+			use_power(5)
+			if(icon_vend) //Show the vending animation if needed
+				flick(icon_vend,src)
+			new R.product_path(drop_location())
+			return TRUE
+		if("eject")
+			if(!coin)
+				to_chat(usr, "<span class='notice'>There is no coin in this machine.</span>")
+				return
+			coin.forceMove(drop_location())
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: remove this line

File: code/game/machinery/vending.dm [hunk 17]
```diff
@@ -605,6 +431,102 @@
...
+							coin = null
+						else
+							to_chat(usr, "<span class='warning'>You couldn't pull [coin] out because your hands are full!</span>")
+							QDEL_NULL(coin)
+					else
+						to_chat(usr, "<span class='warning'>You weren't able to pull [coin] out fast enough, the machine ate it, string and all!</span>")
+						QDEL_NULL(coin)
+				else
+					QDEL_NULL(coin)
+			else if (!(R in product_records))
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: you're missing the checks for coin records and hidden records

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/machinery/vending.dm (hunk 17 above)
Comment (MoreRobustThanYou) [reply code/game/machinery/vending.dm]: that's not needed, as coin/hidden records have already been handled

File: code/game/machinery/vending.dm (hunk 16 above)
Comment (MoreRobustThanYou) [reply code/game/machinery/vending.dm]: which one

**Review by Cyberboss — DISMISSED** [review]

**Comment by Buggy123** [comment]
I like this but I really wish that since TGUI is used by everything, that bug where sometimes it just doesn't load for 20 seconds would be fixed.

**Comment by GunHog** [comment]
I get that with _any_ popup, not just TGUI...

**Comment by Buggy123** [comment]
Not-TGUI is so rare nowadays that it might be the same way for me, but I've just never seen it.

File: code/game/machinery/vending.dm [hunk 18]
```diff
@@ -605,6 +431,103 @@
...
+	if(!vend_ready)
+		return
+	if(issilicon(usr))
+		if(iscyborg(usr))
+			var/mob/living/silicon/robot/R = usr
+			if(!(R.module && istype(R.module,/obj/item/robot_module/butler) ))
+				to_chat("<span class='notice'>[src] refuses to interface with you, as you are not in its target demographic!</span>")
+				return
+		else
+			to_chat(usr, "<span class='notice'>[src] refuses to interface with you, as you are not in its target demographic!</span>")
```
Comment (YPOQ) [inline code/game/machinery/vending.dm]: This was removed. #31541

File: code/game/machinery/vending.dm [hunk 19]
```diff
@@ -350,180 +350,6 @@
...
-
-		if (R.amount <= 0)
-			to_chat(usr, "<span class='warning'>Sold out.</span>")
-			vend_ready = 1
-			return
-		else
-			R.amount--
-
-		if(((last_reply + 200) <= world.time) && vend_reply)
-			speak(vend_reply)
```
Comment (YPOQ) [inline code/game/machinery/vending.dm]: Please port this :)

Comment (MoreRobustThanYou) [reply code/game/machinery/vending.dm]: done

**Comment by Bawhoppen** [comment]
Now we can separate stuff in clothesvend and costume vendor into tabs

**Comment by octareenroon91** [comment]
You're killing me here

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**